### PR TITLE
Support Defining Ordering Equivalence at the Source

### PR DESCRIPTION
**Pull request by berkaysynnada**
# Which issue does this PR close?

Closes #.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/datasource/file_format/options.rs [hunk 1]
```diff
@@ -443,7 +443,7 @@ impl ReadOptions<'_> for CsvReadOptions<'_> {
             .with_target_partitions(config.target_partitions())
             .with_table_partition_cols(self.table_partition_cols.clone())
             // TODO: Add file sort order into CsvReadOptions and introduce here.
-            .with_file_sort_order(None)
```
Comment (mustafasrepo) [inline datafusion/core/src/datasource/file_format/options.rs:446]: I think you shoudln't change `with_file_sort_order api`. Similar to `with order` one can use this method multiple times to define ordering equivalences.

File: datafusion/core/src/physical_plan/mod.rs [hunk 2]
```diff
@@ -578,6 +578,43 @@ impl PartialEq for Partitioning {
...
+        // Return an empty OrderingEquivalenceProperties:
+        return oep;
+    };
+    // First entry among eq_orderings is the head, skip it:
+    for ordering in eq_orderings.iter().skip(1) {
+        let column = ordering
+            .iter()
+            .map(|e| TryFrom::try_from(e.clone()))
+            .collect::<Result<Vec<_>>>();
+        if let Ok(column) = column {
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/mod.rs:609]: Can we safely ignore possible errors here?

File: datafusion/physical-expr/src/equivalence.rs [hunk 3]
```diff
@@ -209,6 +210,24 @@ impl From<OrderedColumn> for PhysicalSortExpr {
     }
 }
 
+impl TryFrom<PhysicalSortExpr> for OrderedColumn {
```
Comment (ozankabak) [inline datafusion/physical-expr/src/equivalence.rs:213]: Is this one of the places we will need to change when we support orderings w.r.t. arbitrary expressions?

**Review by ozankabak — APPROVED** [review]
LGTM, I left two small questions we can discuss in our meeting tomorrow.

I am torn between your original style with API change, and Mustafa's suggestion (i.e. making `with_file_sort_order` applicable more than once). I am currently leaning towards your original style, so I changed it for now. But let's discuss in our meeting tomorrow and try to build a consensus on which one is right.

The obvious advantage of Mustafa's suggestion is that it mirrors the SQL syntax and doesn't require an API change in the common case, which is the main downside of your style. However, Mustafa's suggestion also has some drawbacks I can see:
1. This is not a declarative API like SQL, it is an imperative API and we are introducing a non-idempotent builder-style function call. We can not "undo" a `with_file_sort_order`, or simply set a given end state.
2. A user that currently expects the ordering to "reset" when they call `with_file_sorter_order` multiple times will silently get a wrong behavior.

For these reasons, I currently lean towards an API change, but let's discuss altogether.

**Comment by ozankabak** [comment]
Merged upstream, closing.

**Pull request closed without merging**